state/factory/workingset.go

```diff
@@ -207,6 +207,24 @@ func (ws *workingSet) runAction(
 		return nil, errors.Wrapf(err, "Failed to get hash")
 	}
 	defer ws.ResetSnapshots()
+	sn := ws.Snapshot()
+	defer func() {
```

Comment:
could these two defer functions been merged which will make it easier to read and maintain